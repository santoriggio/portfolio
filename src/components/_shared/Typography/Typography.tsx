import { useAppTheme } from "@/hooks/useAppTheme";
import { SpacingProps, useSpacing } from "@/hooks/useSpacing";
import { TxKeyPath, isRTL, translate } from "@/i18n";
import { ThemedStyle, ThemedStyleArray, colors, typography } from "@/theme";
import { TOptions } from "i18next";
import React, { FC, ReactNode } from "react";
import {
  Text,
  TextProps as RNTextProps,
  TextStyle,
  StyleProp,
} from "react-native";

type Sizes = keyof typeof $sizeStyles;
type Presets = "default" | "bold";
type Colors = keyof typeof colors;

export type TypographyProps = SpacingProps &
  RNTextProps & {
    flex?: boolean;
    /**
     * Text which is looked up via i18n.
     */
    tx?: TxKeyPath;
    /**
     * Optional options to pass to i18n. Useful for interpolation
     * as well as explicitly setting locale or translation fallbacks.
     */
    txOptions?: TOptions;
    /**
     * An optional style override useful for padding & margin.
     */
    style?: StyleProp<TextStyle>;
    /**
     * Children components.
     */
    children?: ReactNode;
    /**
     * Text Size
     */
    size?: Sizes;
    /**
     * Text Color
     */
    color?: Colors;
    /**
     * One of the different types of text presets.
     */
    preset?: Presets;
  };

export const Typography: FC<TypographyProps> = ({
  tx,
  txOptions,
  children,
  size,
  color,
  flex,
  style: $styleOverride,
  ...rest
}) => {
  const { themed, theme } = useAppTheme();
  const $spacing = useSpacing(rest);

  const i18nText = tx && translate(tx, txOptions);
  const content = i18nText || children;
  const preset: Presets = rest.preset ?? "default";

  const $color = color && { color: theme.colors[color] };
  const $flex = flex && { flex: 1 };

  const $styles: StyleProp<TextStyle> = [
    $rtlStyle,
    themed($presets[preset]),
    size && $sizeStyles[size],
    $flex,
    $color,
    $spacing,
    $styleOverride,
  ];

  return (
    <Text {...rest} style={$styles}>
      {content}
    </Text>
  );
};

const $sizeStyles = {
  xxl: { fontSize: 36, lineHeight: 44 } satisfies TextStyle,
  xl: { fontSize: 24, lineHeight: 34 } satisfies TextStyle,
  lg: { fontSize: 20, lineHeight: 32 } satisfies TextStyle,
  md: { fontSize: 18, lineHeight: 26 } satisfies TextStyle,
  sm: { fontSize: 16, lineHeight: 24 } satisfies TextStyle,
  xs: { fontSize: 14, lineHeight: 21 } satisfies TextStyle,
  xxs: { fontSize: 12, lineHeight: 18 } satisfies TextStyle,
};

const $baseStyle: ThemedStyle<TextStyle> = (theme) => ({
  ...$sizeStyles.sm,
  color: theme.colors.text,
});

const $boldStyle: ThemedStyle<TextStyle> = (theme) => ({});

const $presets: Record<Presets, ThemedStyleArray<TextStyle>> = {
  default: [$baseStyle],
  bold: [$baseStyle, $boldStyle],
};
const $rtlStyle: TextStyle = isRTL ? { writingDirection: "rtl" } : {};
