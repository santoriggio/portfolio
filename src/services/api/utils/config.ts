export type SendApiRequestOptions<A extends API> = {
  method?: "GET" | "POST" | "PATCH" | "DELETE";
  endpoint?: string;
  request?: Request[A] | null;
  headers?: Record<string, string>;
  requireAuth?: boolean;
};

export enum API {
  GET_USERS,
  SIGNUP,
  SIGNIN,
  CREATE_STICKER_PACK,
  GET_USER_STICKER_PACK,
  GET_COMMUNITY_SP,
  GET_SP,
  UPLOAD_STICKER,
}

export type Request = {
  [API.GET_USERS]: {
    page?: number;
  };
  [API.SIGNUP]: {
    email: string;
    password: string;
  };
  [API.SIGNIN]: {
    email: string;
    password: string;
  };
  [API.CREATE_STICKER_PACK]: {
    title: string;
  };
  [API.GET_USER_STICKER_PACK]: undefined;

  [API.GET_COMMUNITY_SP]: undefined;
  [API.GET_SP]: {
    id: number;
  };
  [API.UPLOAD_STICKER]: {
    base64: string;
  };
};

export type Response = {
  [API.GET_USERS]: undefined;
  [API.SIGNUP]: {
    id: number;
    message: string;
    accessToken: string;
    email: string;
    username: string;
  };
  [API.SIGNIN]: {
    id: number;
    message: string;
    accessToken: string;
    email: string;
    username: string;
  };
  [API.CREATE_STICKER_PACK]: {
    id: number;
    users: any[];
  };
  [API.GET_USER_STICKER_PACK]: {
    list: any[];
  };
  [API.GET_COMMUNITY_SP]: {
    list: any[];
  };
  [API.GET_SP]: {
    pack: any;
  };
  [API.UPLOAD_STICKER]: {
    id: number;
  };
};

export const optionsMapping: Record<API, SendApiRequestOptions<any>> = {
  [API.GET_USERS]: {
    endpoint: "/users",
  },
  [API.SIGNUP]: {
    endpoint: "/auth/register",
    method: "POST",
  },
  [API.SIGNIN]: {
    endpoint: "/auth/login",
    method: "POST",
  },
  [API.CREATE_STICKER_PACK]: {
    endpoint: "/sticker-pack/create",
    method: "POST",
  },
  [API.GET_USER_STICKER_PACK]: {
    endpoint: "/sticker-pack/user",
    method: "GET",
  },
  [API.GET_COMMUNITY_SP]: {
    endpoint: "/sticker-pack/community",
    method: "GET",
  },
  [API.GET_SP]: {
    endpoint: "/sticker-pack/",
    method: "POST",
  },
  [API.UPLOAD_STICKER]: {
    endpoint: "/sticker/upload",
    method: "POST",
  },
};

export type State = {
  [API.GET_USERS]: {
    list: number;
  };
  [API.SIGNUP]: {
    id: number;
    email: string;
  };
  [API.SIGNIN]: {
    id: number;
    email: string;
  };
  [API.CREATE_STICKER_PACK]: {
    pack: {
      id: number;
      users: any[];
    };
  };
  [API.GET_USER_STICKER_PACK]: {
    list: any[];
  };
  [API.GET_COMMUNITY_SP]: {
    list: any[];
  };
  [API.GET_SP]: {
    pack: any;
  };
  [API.UPLOAD_STICKER]: {
    id: number;
  };
};

export type Store<A extends API> = {
  loading: boolean;
  error: Error | null;
  request: Request[A] | null;
  state: State[A] | null;
};

export type SendApiRequestReturn<A extends API> = {
  res: Response[A] | null;
  err: Error | null;
};
