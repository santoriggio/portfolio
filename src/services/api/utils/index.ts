export * from "./createApiHandler";
export * from "./sendApiRequest";
