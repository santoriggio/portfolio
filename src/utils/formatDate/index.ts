export * from './formatDate'
export * from './formatDate.types'
