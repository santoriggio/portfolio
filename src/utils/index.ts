export * from "./setAccessTokenAsync";
export * from "./getAccessTokenAsync";
