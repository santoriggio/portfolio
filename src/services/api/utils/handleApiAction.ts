import { StoreApi, UseBoundStore } from "zustand";
import { API, Store, Response } from "./config";

function handleApiAction<A extends API>(
  type: "error",
  callback: (store: UseBoundStore<StoreApi<Store<A>>>, err: Error) => void,
): {
  type: "error";
  callback: (store: UseBoundStore<StoreApi<Store<A>>>, err: Error) => void;
};

function handleApiAction<A extends API>(
  type: "success",
  callback: (
    store: UseBoundStore<StoreApi<Store<A>>>,
    res: Response[A],
  ) => void,
): {
  type: "success";
  callback: (
    store: UseBoundStore<StoreApi<Store<A>>>,
    res: Response[A],
  ) => void;
};

function handleApiAction<A extends API>(
  type: "success" | "error",
  callback:
    | ((store: UseBoundStore<StoreApi<Store<A>>>, err: Error) => void)
    | ((store: UseBoundStore<StoreApi<Store<A>>>, res: Response[A]) => void),
): {
  type: "success" | "error";
  callback:
    | ((store: UseBoundStore<StoreApi<Store<A>>>, err: Error) => void)
    | ((store: UseBoundStore<StoreApi<Store<A>>>, res: Response[A]) => void);
} {
  return { type, callback };
}

export { handleApiAction };
