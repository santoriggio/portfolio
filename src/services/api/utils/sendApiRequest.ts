import { getAccessTokenAsync } from "@/utils";
import {
  API,
  SendApiRequestOptions,
  SendApiRequestReturn,
  optionsMapping,
} from "./config";

export const sendApiRequest = async <A extends API>(
  api: A,
  options: SendApiRequestOptions<A>,
): Promise<SendApiRequestReturn<A>> => {
  const customOptions: SendApiRequestOptions<A> = {
    requireAuth: true,
    ...optionsMapping[api],
    ...options,
  };

  if (!customOptions.endpoint) {
    return {
      res: null,
      err: null,
    };
  }

  try {
    const url = "http://localhost:3000" + customOptions.endpoint;

    const access_token = await getAccessTokenAsync();

    const headers: Record<string, string> = {
      Accept: "application/json",
      Authorization: `Bearer ${access_token}`,
      "Content-Type": "application/json",
      ...customOptions.headers,
    };

    if (!customOptions.requireAuth) {
      delete headers.Authorization;
    }

    const res = await fetch(url, {
      method: customOptions.method,
      headers,
      body: customOptions.request && JSON.stringify(customOptions.request),
    });


    if (res.status === 401) {
      throw "no_access";
    }

    const json = await res.json();

    if (json.error && json.message) {
      console.log(json.error);
      throw Error(json.message);
    }

    return { res: json, err: null };
  } catch (err: any) {
    return { res: null, err };
  }
};
