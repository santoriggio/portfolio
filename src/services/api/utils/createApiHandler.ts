import { StoreApi, UseBoundStore, create } from "zustand";
import { API, Request, Response, Store } from "./config";

import { sendApiRequest } from "./sendApiRequest";

type SuccessAction<A extends API> = {
  type: "success";
  callback: (
    store: UseBoundStore<StoreApi<Store<A>>>,
    res: Response[A],
  ) => void;
};
type ErrorAction<A extends API> = {
  type: "error";
  callback: (store: UseBoundStore<StoreApi<Store<A>>>, err: Error) => void;
};

type Action<A extends API> = SuccessAction<A> | ErrorAction<A>;

export const createApiHandler = <A extends API>(
  api: A,
  ...actions: Action<A>[]
) => {
  const controller = create<Store<A>>(() => ({
    state: null,
    request: null,
    loading: false,
    error: null,
  }));

  return {
    useStore: controller,
    request: (request: Request[A]) => {
      controller.setState({ request });
    },
    execute: async () => {
      try {
        controller.setState({ loading: true });
        const { request } = controller.getState();
        const { res, err } = await sendApiRequest(api, { request });

        if (err) {
          throw err;
        }

        controller.setState({ error: null });

        for (const action of actions) {
          if (action.type === "success" && res) {
            action.callback(controller, res);
          }
        }
      } catch (error: any) {
        if (typeof error === "string") {
          controller.setState({ error: new Error(error) });
        } else if (error instanceof Error) {
          controller.setState({ error });
        }

        for (const action of actions) {
          if (action.type === "error") {
            action.callback(controller, error);
          }
        }
      } finally {
        controller.setState({ loading: false });
      }
    },
  };
};
