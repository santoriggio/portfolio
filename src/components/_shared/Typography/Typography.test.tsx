import { render } from "@/utils";
import { Typography } from "./Typography";

describe("<Typography />", () => {
  it("renders correcly", () => {
    const tree = render(<Typography />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("has text", () => {
    const { getByText } = render(<Typography>Test</Typography>);
    expect(getByText("Test")).toBeTruthy();
  });
});
