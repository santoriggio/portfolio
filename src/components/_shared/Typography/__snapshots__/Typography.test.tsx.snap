// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography/> renders correcly 1`] = `
<Text
  style={
    [
      {},
      {
        "color": "#333333",
        "fontFamily": "spaceGroteskRegular",
        "fontSize": 16,
        "lineHeight": 24,
      },
      undefined,
      undefined,
      undefined,
      {},
      undefined,
    ]
  }
/>
`;
